File: backend/app/schemas/voice.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel

from ..models.task import TaskPriority, TaskStatus
from .task import TaskRead


class VoiceTaskCandidate(BaseModel):
    id: int
    title: str
    due_date: Optional[datetime] = None
    start_date: Optional[datetime] = None
    status: TaskStatus
    priority: TaskPriority


class VoiceTaskDraft(BaseModel):
    title: str
    description: Optional[str] = None
    start_date: Optional[datetime] = None
    due_date: Optional[datetime] = None
    status: TaskStatus = TaskStatus.todo
    priority: TaskPriority = TaskPriority.medium
    is_recurring: bool = False
    recurring_rule: Optional[str] = None
    remind_before: Optional[int] = None
    category_id: Optional[int] = None
    role_id: Optional[int] = None


class VoiceTaskResponse(BaseModel):
    transcript: str
    task: Optional[TaskRead] = None
    action: Optional[str] = None
    matched: Optional[bool] = None
    status: Optional[str] = None
    candidates: Optional[List[VoiceTaskCandidate]] = None
    draft: Optional[VoiceTaskDraft] = None
    suggested_task: Optional[VoiceTaskCandidate] = None

File: frontend/web/src/services/voice.ts
import { api } from './apiClient';
import { Task, TaskPriority, TaskStatus } from './tasks';

export interface VoiceTaskResponse {
  transcript: string;
  task?: Task;
  action?: string | null;
  matched?: boolean | null;
  status?: TaskStatus | null;
  candidates?: VoiceTaskCandidate[] | null;
  draft?: VoiceTaskDraft | null;
  suggested_task?: VoiceTaskCandidate | null;
}

export interface VoiceTaskCandidate {
  id: number;
  title: string;
  due_date?: string | null;
  start_date?: string | null;
  status: TaskStatus;
  priority: 'high' | 'medium' | 'low';
}

export interface VoiceTaskDraft {
  title: string;
  description?: string | null;
  start_date?: string | null;
  due_date?: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  is_recurring?: boolean;
  recurring_rule?: string | null;
  remind_before?: number | null;
  category_id?: number | null;
  role_id?: number | null;
}

export async function createTaskFromVoice(audio: Blob): Promise<VoiceTaskResponse> {
  const form = new FormData();
  const type = audio.type || '';
  const filename = type.includes('wav') ? 'voice.wav' : 'voice.ogg';
  form.append('audio', audio, filename);
  const { data } = await api.post<VoiceTaskResponse>('/voice/tasks', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}
